python/triton/compiler.py

```diff
@@ -1546,31 +1561,33 @@ def format_of(ty):
   return ptr_info;
 }}
 
-static PyObject* launch(PyObject* self, PyObject* args) {{
-  int gridX, gridY, gridZ;
-  uint64_t _stream;
-  uint64_t _function;
-  int num_warps;
-  int shared_memory;
-  PyObject *launch_enter_hook = NULL;
-  PyObject *launch_exit_hook = NULL;
-  PyObject *compiled_kernel = NULL;
-  {' '.join([f"{_extracted_type(ty)} _arg{i}; " for i, ty in signature.items()])}
-  if(!PyArg_ParseTuple(args, \"{format}\", &gridX, &gridY, &gridZ, &num_warps, &shared_memory, &_stream, &_function, &launch_enter_hook, &launch_exit_hook, &compiled_kernel, {', '.join(f"&_arg{i}" for i, ty in signature.items())})) {{
-    return NULL;
-  }}
+static PyObject* launch(PyObject* self, PyObject* const* args, Py_ssize_t nargs) {{
+  int gridX = (int) PyLong_AsLong(args[0]);
+  int gridY = (int) PyLong_AsLong(args[1]);
+  int gridZ = (int) PyLong_AsLong(args[2]);
+  int num_warps = (int) PyLong_AsLong(args[3]);
+  int shared_memory = (int) PyLong_AsLong(args[4]);
+
+  uint64_t _stream = PyLong_AsUnsignedLongLong(args[5]);
+  uint64_t _function = PyLong_AsUnsignedLongLong(args[6]);
+
+  PyObject *launch_enter_hook = args[7];
+  PyObject *launch_exit_hook = args[8];
+  PyObject *compiled_kernel = args[9];
+
+  {" ".join([f"PyObject* _arg{i} = args[{10+i}];" if ty[0] == "*" else "" for i, ty in signature.items()])};
+  {" ".join([f"{ty_to_cpp(ty)} _arg{i} = ({ty_to_cpp(ty)}) {ty_to_cast_method(ty)}(args[{10+i}]);" if ty[0] != "*" else "" for i, ty in signature.items()])};
 
   if (launch_enter_hook != Py_None) {{
-    PyObject_CallObject(launch_enter_hook, args);
+    PyObject_CallObject(launch_enter_hook, *args);
```

Comment:
This is clearly wrong since `*args` is the same as `args[0]` which is the `PyObject*` for `gridX` and not a tuple of arguments.

Reply:
Thanks for the good catch!

Do you think we could use something like `PyObject_Vectorcall` to avoid creating another PyObject?

Reply:
Unfortunately `PyObject_Vectorcall` was not public until python 3.8 (and you had to call it as `_PyObject_Vectorcall`  until 3.9).